File: apps/nomadically.work/workers/eu-classifier/src/signals.py
"""Deterministic signal extraction from ATS-enriched job data.

Extracts structured boolean/string signals (remote flags, country codes,
negative signals, EU timezone mentions) that feed both the keyword
heuristic and the LLM classification prompt.
"""

import json
import re

from constants import (
    EU_ISO_CODES,
    EU_COUNTRY_NAMES,
    COUNTRY_NAME_TO_ISO,
    NEGATIVE_EU_PATTERN,
    US_IMPLICIT_PATTERN,
    EU_TIMEZONE_PATTERN,
    normalize_text_for_signals,
)


def extract_eu_signals(job: dict) -> dict:
    """Extract deterministic EU-related signals from ATS-enriched job data.

    Returns a dict of boolean/string signals that the keyword heuristic
    and LLM prompt can use for classification.
    """
    signals: dict = {
        "ats_remote": False,
        "eu_country_code": False,
        "country_code": None,
        "negative_signals": [],
        "us_implicit_signals": [],
        "eu_timezone": False,
        "eu_countries_in_location": [],
        "all_locations": [],
    }

    # ATS remote flag -- from ashby_is_remote or workplace_type
    ashby_remote = job.get("ashby_is_remote")
    workplace = (job.get("workplace_type") or "").lower()
    location_lower = (job.get("location") or "").lower().strip()
    if (ashby_remote == 1 or ashby_remote is True
            or workplace == "remote"
            or location_lower == "remote"
            or location_lower.startswith("remote ")):
        signals["ats_remote"] = True

    # Country code -> EU membership check
    raw_country = (job.get("country") or "").strip()
    country = raw_country.upper()
    if country and re.fullmatch(r"[A-Z]{2,3}", country):
        signals["country_code"] = country
        if country in EU_ISO_CODES:
            signals["eu_country_code"] = True
    elif raw_country:
        iso = COUNTRY_NAME_TO_ISO.get(raw_country.lower())
        if iso:
            signals["country_code"] = iso
            if iso in EU_ISO_CODES:
                signals["eu_country_code"] = True

    # Fallback: extract country from ashby_address if still no country_code
    if not signals["country_code"]:
        try:
            addr = job.get("ashby_address")
            if isinstance(addr, str):
                addr = json.loads(addr)
            if isinstance(addr, dict):
                postal = addr.get("postalAddress") or addr
                addr_country = (postal.get("addressCountry") or "").strip()
                addr_locality = (postal.get("addressLocality") or "").strip()
                for candidate in [addr_country, addr_locality]:
                    if not candidate:
                        continue
                    upper = candidate.upper()
                    if re.fullmatch(r"[A-Z]{2,3}", upper):
                        signals["country_code"] = upper
                        if upper in EU_ISO_CODES:
                            signals["eu_country_code"] = True
                        break
                    iso = COUNTRY_NAME_TO_ISO.get(candidate.lower())
                    if iso:
                        signals["country_code"] = iso
                        if iso in EU_ISO_CODES:
                            signals["eu_country_code"] = True
                        break
        except Exception:
            pass

    # Fallback: extract country from location string (e.g. "USA | Remote")
    if not signals["country_code"] and location_lower:
        for token in re.split(r"[|,/()\-\u2013\u2014]+", location_lower):
            token = token.strip().rstrip(".")
            if not token or token == "remote":
                continue
            iso = COUNTRY_NAME_TO_ISO.get(token)
            if iso:
                signals["country_code"] = iso
                if iso in EU_ISO_CODES:
                    signals["eu_country_code"] = True
                break
            upper = token.upper()
            if re.fullmatch(r"[A-Z]{2}", upper):
                signals["country_code"] = upper
                if upper in EU_ISO_CODES:
                    signals["eu_country_code"] = True
                break

    # Negative signals via regex on description
    desc = (job.get("description") or "")[:8000].lower()
    location = (job.get("location") or "").lower()
    full_text = normalize_text_for_signals(f"{location} {desc}")
    for m in NEGATIVE_EU_PATTERN.finditer(full_text):
        signals["negative_signals"].append(m.group(0))

    # US-implicit signals via regex on description
    us_implicit: list[str] = []
    for m in US_IMPLICIT_PATTERN.finditer(full_text):
        us_implicit.append(m.group(0))
    signals["us_implicit_signals"] = us_implicit

    # EU timezone / business hours
    if EU_TIMEZONE_PATTERN.search(full_text):
        signals["eu_timezone"] = True

    # EU country names in location string
    for name in EU_COUNTRY_NAMES:
        if name in location:
            signals["eu_countries_in_location"].append(name)

    # Aggregate all ATS locations
    all_locs: list[str] = []
    if job.get("location"):
        all_locs.append(job["location"])

    # offices (Greenhouse JSON array)
    try:
        offices = job.get("offices")
        if isinstance(offices, str):
            offices = json.loads(offices)
        if isinstance(offices, list):
            for o in offices:
                name = o.get("name") or o.get("location") if isinstance(o, dict) else str(o)
                if name:
                    all_locs.append(name)
    except Exception:
        pass

    # categories.allLocations (Ashby/Lever JSON)
    try:
        cats = job.get("categories")
        if isinstance(cats, str):
            cats = json.loads(cats)
        if isinstance(cats, dict):
            for loc in (cats.get("allLocations") or []):
                if loc and loc not in all_locs:
                    all_locs.append(loc)
    except Exception:
        pass

    # ashby_secondary_locations
    try:
        sec = job.get("ashby_secondary_locations")
        if isinstance(sec, str):
            sec = json.loads(sec)
        if isinstance(sec, list):
            for s in sec:
                loc_name = s.get("location") if isinstance(s, dict) else str(s)
                if loc_name and loc_name not in all_locs:
                    all_locs.append(loc_name)
    except Exception:
        pass

    signals["all_locations"] = all_locs
    return signals


def format_signals(signals: dict) -> str:
    """Format extracted signals as a text block for the LLM prompt."""
    parts: list[str] = []

    if signals["ats_remote"]:
        parts.append("- ATS remote flag: YES")
    if signals["country_code"]:
        eu_label = " (EU member)" if signals["eu_country_code"] else " (NOT EU)"
        parts.append(f"- Country code: {signals['country_code']}{eu_label}")
    if signals["negative_signals"]:
        parts.append(f"- Negative signals: {', '.join(signals['negative_signals'][:5])}")
    if signals["eu_timezone"]:
        parts.append("- EU timezone/business hours signal detected")
    if signals["eu_countries_in_location"]:
        parts.append(f"- EU countries in location: {', '.join(signals['eu_countries_in_location'][:5])}")
    if len(signals["all_locations"]) > 1:
        parts.append(f"- All ATS locations: {', '.join(signals['all_locations'][:8])}")

    return "\n".join(parts) if parts else "- No structured ATS signals available"

File: apps/lead-gen/crates/nomad/src/signals.rs
use regex::Regex;

use crate::constants::{
    COUNTRY_NAME_TO_ISO, EU_COUNTRY_NAMES, EU_ISO_CODES, EU_TIMEZONE_PATTERN,
    NEGATIVE_EU_PATTERN, US_IMPLICIT_PATTERN, normalize_text_for_signals,
};
use crate::{JobRow, SignalSet};

/// Extract deterministic EU-related signals from ATS-enriched job data.
pub fn extract_eu_signals(job: &JobRow) -> SignalSet {
    let mut signals = SignalSet::default();

    // ATS remote flag
    let workplace = job.get_str("workplace_type").to_lowercase();
    let location_lower = job.get_str("location").to_lowercase().trim().to_string();

    if job.ashby_is_remote_bool()
        || workplace == "remote"
        || location_lower == "remote"
        || location_lower.starts_with("remote ")
    {
        signals.ats_remote = true;
    }

    // Country code -> EU membership check
    let raw_country = job.get_str("country").trim().to_string();
    let country_upper = raw_country.to_uppercase();

    if !country_upper.is_empty() && is_iso_code(&country_upper) {
        signals.country_code = Some(country_upper.clone());
        if EU_ISO_CODES.contains(country_upper.as_str()) {
            signals.eu_country_code = true;
        }
    } else if !raw_country.is_empty() {
        if let Some(&iso) = COUNTRY_NAME_TO_ISO.get(raw_country.to_lowercase().as_str()) {
            signals.country_code = Some(iso.to_string());
            if EU_ISO_CODES.contains(iso) {
                signals.eu_country_code = true;
            }
        }
    }

    // Fallback: extract country from ashby_address
    if signals.country_code.is_none() {
        if let Some(ref addr_str) = job.ashby_address {
            if let Ok(addr) = serde_json::from_str::<serde_json::Value>(addr_str) {
                let postal = addr.get("postalAddress").unwrap_or(&addr);
                for key in &["addressCountry", "addressLocality"] {
                    if let Some(candidate) = postal.get(key).and_then(|v| v.as_str()) {
                        let candidate = candidate.trim();
                        if candidate.is_empty() {
                            continue;
                        }
                        let upper = candidate.to_uppercase();
                        if is_iso_code(&upper) {
                            signals.country_code = Some(upper.clone());
                            if EU_ISO_CODES.contains(upper.as_str()) {
                                signals.eu_country_code = true;
                            }
                            break;
                        }
                        if let Some(&iso) =
                            COUNTRY_NAME_TO_ISO.get(candidate.to_lowercase().as_str())
                        {
                            signals.country_code = Some(iso.to_string());
                            if EU_ISO_CODES.contains(iso) {
                                signals.eu_country_code = true;
                            }
                            break;
                        }
                    }
                }
            }
        }
    }

    // Fallback: extract country from location string (e.g. "USA | Remote")
    if signals.country_code.is_none() && !location_lower.is_empty() {
        let split_re = Regex::new(r"[|,/()\-\u{2013}\u{2014}]+").unwrap();
        for token in split_re.split(&location_lower) {
            let token = token.trim().trim_end_matches('.');
            if token.is_empty() || token == "remote" {
                continue;
            }
            if let Some(&iso) = COUNTRY_NAME_TO_ISO.get(token) {
                signals.country_code = Some(iso.to_string());
                if EU_ISO_CODES.contains(iso) {
                    signals.eu_country_code = true;
                }
                break;
            }
            let upper = token.to_uppercase();
            if upper.len() == 2 && upper.chars().all(|c| c.is_ascii_alphabetic()) {
                signals.country_code = Some(upper.clone());
                if EU_ISO_CODES.contains(upper.as_str()) {
                    signals.eu_country_code = true;
                }
                break;
            }
        }
    }

    // Negative signals via regex on description
    let desc = &job.get_str("description")[..job.get_str("description").len().min(8000)];
    let location = job.get_str("location").to_lowercase();
    let full_text = normalize_text_for_signals(&format!("{} {}", location, desc.to_lowercase()));

    for m in NEGATIVE_EU_PATTERN.find_iter(&full_text) {
        signals.negative_signals.push(m.as_str().to_string());
    }

    // US-implicit signals
    for m in US_IMPLICIT_PATTERN.find_iter(&full_text) {
        signals.us_implicit_signals.push(m.as_str().to_string());
    }

    // EU timezone / business hours
    if EU_TIMEZONE_PATTERN.is_match(&full_text) {
        signals.eu_timezone = true;
    }

    // EU country names in location string
    for name in EU_COUNTRY_NAMES.iter() {
        if location.contains(name) {
            signals.eu_countries_in_location.push(name.to_string());
        }
    }

    // Aggregate all ATS locations
    let mut all_locs: Vec<String> = Vec::new();
    if !job.get_str("location").is_empty() {
        all_locs.push(job.get_str("location").to_string());
    }

    // offices (Greenhouse JSON array)
    if let Some(ref offices_str) = job.offices {
        if let Ok(offices) = serde_json::from_str::<serde_json::Value>(offices_str) {
            if let Some(arr) = offices.as_array() {
                for o in arr {
                    let name = o
                        .get("name")
                        .or_else(|| o.get("location"))
                        .and_then(|v| v.as_str());
                    if let Some(name) = name {
                        all_locs.push(name.to_string());
                    }
                }
            }
        }
    }

    // categories.allLocations (Ashby/Lever JSON)
    if let Some(ref cats_str) = job.categories {
        if let Ok(cats) = serde_json::from_str::<serde_json::Value>(cats_str) {
            if let Some(locs) = cats.get("allLocations").and_then(|v| v.as_array()) {
                for loc in locs {
                    if let Some(s) = loc.as_str() {
                        if !s.is_empty() && !all_locs.contains(&s.to_string()) {
                            all_locs.push(s.to_string());
                        }
                    }
                }
            }
        }
    }

    // ashby_secondary_locations
    if let Some(ref sec_str) = job.ashby_secondary_locations {
        if let Ok(sec) = serde_json::from_str::<serde_json::Value>(sec_str) {
            if let Some(arr) = sec.as_array() {
                for s in arr {
                    let loc_name = s
                        .get("location")
                        .and_then(|v| v.as_str())
                        .or_else(|| s.as_str());
                    if let Some(name) = loc_name {
                        if !name.is_empty() && !all_locs.contains(&name.to_string()) {
                            all_locs.push(name.to_string());
                        }
                    }
                }
            }
        }
    }

    signals.all_locations = all_locs;
    signals
}

/// Format extracted signals as a text block for the LLM prompt.
pub fn format_signals(signals: &SignalSet) -> String {
    let mut parts: Vec<String> = Vec::new();

    if signals.ats_remote {
        parts.push("- ATS remote flag: YES".to_string());
    }
    if let Some(ref cc) = signals.country_code {
        let eu_label = if signals.eu_country_code {
            " (EU member)"
        } else {
            " (NOT EU)"
        };
        parts.push(format!("- Country code: {cc}{eu_label}"));
    }
    if !signals.negative_signals.is_empty() {
        let joined = signals.negative_signals[..signals.negative_signals.len().min(5)].join(", ");
        parts.push(format!("- Negative signals: {joined}"));
    }
    if signals.eu_timezone {
        parts.push("- EU timezone/business hours signal detected".to_string());
    }
    if !signals.eu_countries_in_location.is_empty() {
        let joined =
            signals.eu_countries_in_location[..signals.eu_countries_in_location.len().min(5)]
                .join(", ");
        parts.push(format!("- EU countries in location: {joined}"));
    }
    if signals.all_locations.len() > 1 {
        let joined = signals.all_locations[..signals.all_locations.len().min(8)].join(", ");
        parts.push(format!("- All ATS locations: {joined}"));
    }

    if parts.is_empty() {
        "- No structured ATS signals available".to_string()
    } else {
        parts.join("\n")
    }
}

fn is_iso_code(s: &str) -> bool {
    (s.len() == 2 || s.len() == 3) && s.chars().all(|c| c.is_ascii_alphabetic())
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn extracts_eu_country_from_country_field() {
        let job = JobRow {
            id: 1,
            country: Some("DE".to_string()),
            ..Default::default()
        };
        let signals = extract_eu_signals(&job);
        assert!(signals.eu_country_code);
        assert_eq!(signals.country_code, Some("DE".to_string()));
    }

    #[test]
    fn extracts_remote_from_ashby() {
        let job = JobRow {
            id: 2,
            ashby_is_remote: Some(serde_json::Value::Bool(true)),
            ..Default::default()
        };
        let signals = extract_eu_signals(&job);
        assert!(signals.ats_remote);
    }

    #[test]
    fn extracts_us_country_from_name() {
        let job = JobRow {
            id: 3,
            country: Some("united states".to_string()),
            ..Default::default()
        };
        let signals = extract_eu_signals(&job);
        assert!(!signals.eu_country_code);
        assert_eq!(signals.country_code, Some("US".to_string()));
    }

    #[test]
    fn detects_negative_signals() {
        let job = JobRow {
            id: 4,
            description: Some("This role is US only and requires US work authorization".to_string()),
            ..Default::default()
        };
        let signals = extract_eu_signals(&job);
        assert!(!signals.negative_signals.is_empty());
    }

    #[test]
    fn detects_eu_timezone() {
        let job = JobRow {
            id: 5,
            description: Some("Must work in EU timezone".to_string()),
            ..Default::default()
        };
        let signals = extract_eu_signals(&job);
        assert!(signals.eu_timezone);
    }

    #[test]
    fn extracts_country_from_location_fallback() {
        let job = JobRow {
            id: 6,
            location: Some("USA | Remote".to_string()),
            ..Default::default()
        };
        let signals = extract_eu_signals(&job);
        assert_eq!(signals.country_code, Some("US".to_string()));
        assert!(!signals.eu_country_code);
    }

    #[test]
    fn format_signals_empty() {
        let signals = SignalSet::default();
        assert_eq!(format_signals(&signals), "- No structured ATS signals available");
    }

    #[test]
    fn format_signals_full() {
        let signals = SignalSet {
            ats_remote: true,
            eu_country_code: true,
            country_code: Some("DE".to_string()),
            eu_timezone: true,
            ..Default::default()
        };
        let text = format_signals(&signals);
        assert!(text.contains("ATS remote flag: YES"));
        assert!(text.contains("DE (EU member)"));
        assert!(text.contains("EU timezone"));
    }
}
